File: Python/DesignPatterns/Abstract_FactoryDesign.py
from abc import ABC, abstractmethod


# Products in UI
class IButton(ABC):
    @abstractmethod
    def render_button(self) -> None:
        raise NotImplementedError


class IModal(ABC):
    @abstractmethod
    def render_modal(self) -> None:
        raise NotImplementedError


class IScreen(ABC):
    @abstractmethod
    def render_screen(self) -> None:
        raise NotImplementedError


# Windows
class WindowButton(IButton):
    def render_button(self) -> None:
        print("Window button rendered")


class WindowModal(IModal):
    def render_modal(self) -> None:
        print("Window modal rendered")


class WindowScreen(IScreen):
    def render_screen(self) -> None:
        print("Window screen rendered")


# Mac
class MacButton(IButton):
    def render_button(self) -> None:
        print("Mac button rendered")


class MacModal(IModal):
    def render_modal(self) -> None:
        print("Mac modal rendered")


class MacScreen(IScreen):
    def render_screen(self) -> None:
        print("Mac screen rendered")


# Linux
class LinuxButton(IButton):
    def render_button(self) -> None:
        print("linux button rendered")


class LinuxModal(IModal):
    def render_modal(self) -> None:
        print("linux modal rendered")


class LinuxScreen(IScreen):
    def render_screen(self) -> None:
        print("linux screen rendered")


class IUIFactory(ABC):
    @abstractmethod
    def get_button(self) -> IButton:
        raise NotImplementedError

    @abstractmethod
    def get_modal(self) -> IModal:
        raise NotImplementedError

    @abstractmethod
    def get_screen(self) -> IScreen:
        raise NotImplementedError


class LinuxFactory(IUIFactory):
    def get_button(self) -> IButton:
        return LinuxButton()

    def get_modal(self) -> IModal:
        return LinuxModal()

    def get_screen(self) -> IScreen:
        return LinuxScreen()


class WindowUIFactory(IUIFactory):
    def get_button(self) -> IButton:
        return WindowButton()

    def get_modal(self) -> IModal:
        return WindowModal()

    def get_screen(self) -> IScreen:
        return WindowScreen()


class MacUIFactory(IUIFactory):
    def get_button(self) -> IButton:
        return MacButton()

    def get_modal(self) -> IModal:
        return MacModal()

    def get_screen(self) -> IScreen:
        return MacScreen()


class UIRender:
    def __init__(self, factory: IUIFactory) -> None:
        self.button = factory.get_button()
        self.modal = factory.get_modal()
        self.screen = factory.get_screen()
        self.render_ui()

    def render_ui(self) -> None:
        self.button.render_button()
        self.modal.render_modal()
        self.screen.render_screen()

    def toggle_ui(self, factory: IUIFactory) -> None:
        self.button = factory.get_button()
        self.modal = factory.get_modal()
        self.screen = factory.get_screen()
        self.render_ui()

    def toggle(self, factory: IUIFactory) -> None:
        self.toggle_ui(factory)


def main() -> None:
    UIRender(LinuxFactory())


if __name__ == "__main__":
    main()

File: Java/DesignPatterns/Abstract_FactoryDesign.java


//Products in UI

interface IButton{
    void renderButton();
}

interface IModal{
    void renderModal();
}

interface IScreen{
    void renderScreen();
}

//wINDOW MAC, LINUX





















//Windows 
class WindowButton implements IButton{
    @Override
    public void renderButton() {
        System.out.println("Window button rendered");
    }
}

class WindowModal implements IModal{

    @Override
    public void renderModal() {
        System.out.println("Window modal rendered");
    }  
}

class WindowScreen implements IScreen{
    @Override
    public void renderScreen() {
        System.out.println("Window screen rendered");
    }    
}

//Mac 
class MacButton implements IButton{
    @Override
    public void renderButton() {
        System.out.println("Mac button rendered");
    }
}

class MacModal implements IModal{

    @Override
    public void renderModal() {
        System.out.println("Mac modal rendered");
    }  
}

class MacScreen implements IScreen{
    @Override
    public void renderScreen() {
        System.out.println("Mac screen rendered");
    }    
}

//Linux

//Mac 
class LinuxButton implements IButton{
    @Override
    public void renderButton() {
        System.out.println("linux button rendered");
    }
}

class LinuxModal implements IModal{

    @Override
    public void renderModal() {
        System.out.println("linux modal rendered");
    }  
}

class LinuxScreen implements IScreen{
    @Override
    public void renderScreen() {
        System.out.println("linux screen rendered");
    }    
}



interface I_UI_Factory{
    IButton getButton();
    IModal getModal();
    IScreen getScreen();
}

class LinuxFactory implements I_UI_Factory{

    @Override
    public IButton getButton() {
        return new LinuxButton();
    }

    @Override
    public IModal getModal() {
        return new LinuxModal();
    }

    @Override
    public IScreen getScreen() {
        return new LinuxScreen();
    }
    
}

class WindowUIFactory implements I_UI_Factory{
    @Override
    public IButton getButton() {
        return new WindowButton();
    }

    @Override
    public IModal getModal() {
        return new WindowModal();
    }

    @Override
    public IScreen getScreen() {
        return new WindowScreen();
    }
}


class MacUIFactory implements I_UI_Factory{
    @Override
    public IButton getButton() {
        return new MacButton();
    }

    @Override
    public IModal getModal() {
        return new MacModal();
    }

    @Override
    public IScreen getScreen() {
        return new MacScreen();
    }
}











class UIRender{
    IButton button;
    IModal modal;
    IScreen screen;

    public UIRender(I_UI_Factory factory){
        this.button = factory.getButton();
        this.modal = factory.getModal();
        this.screen = factory.getScreen();
        renderUI();
    }

    void renderUI(){
        button.renderButton();
        modal.renderModal();
        screen.renderScreen();
    }

    void toggleUI(I_UI_Factory factory){
        this.button = factory.getButton();
        this.modal = factory.getModal();
        this.screen = factory.getScreen();
        renderUI();
    }


    void toggle(I_UI_Factory factory){
        toggleUI(factory);
    }
}




public class Abstract_FactoryDesign {
    public static void main(String[] args) {
        UIRender myView = new UIRender(new LinuxFactory());

    }
}
